src/uptane/partialverificationsecondary.cc

```diff
@@ -0,0 +1,100 @@
+#include "uptane/partialverificationsecondary.h"
+
+#include <string>
+#include <vector>
+
+#include <boost/filesystem.hpp>
+#include "json/json.h"
+
+#include "../exceptions.h"
+#include "types.h"
+#include "uptane/secondaryconfig.h"
+#include "uptane/secondaryinterface.h"
+#include "uptane/tufrepository.h"
+
+namespace Uptane {
+
+PartialVerificationSecondary::PartialVerificationSecondary(const SecondaryConfig &sconfig_in)
+    : SecondaryInterface(sconfig_in), root_(Root::kAcceptAll) {
+  boost::filesystem::create_directories(sconfig.metadata_path);
+
+  // FIXME Probably we need to generate keys on the secondary
+  if (!loadKeys(&public_key_, &private_key_)) {
+    if (!Crypto::generateKeyPair(sconfig.key_type, &public_key_, &private_key_)) {
+      LOG_ERROR << "Could not generate keys for secondary " << PartialVerificationSecondary::getSerial() << "@"
+                << sconfig.ecu_hardware_id;
+      throw std::runtime_error("Unable to generate secondary keys");
+    }
+    storeKeys(public_key_, private_key_);
+  }
+  public_key_id_ = Crypto::getKeyId(public_key_);
+}
+
+bool PartialVerificationSecondary::putMetadata(const MetaPack &meta) {
+  TimeStamp now(TimeStamp::Now());
+  detected_attack_.clear();
+
+  root_.UnpackSignedObject(now, "director", Role::Root(), meta.director_root.original());
+  root_ = meta.director_root;
+  root_.UnpackSignedObject(now, "director", Role::Targets(), meta.director_targets.original());
```

Comment:
I'm worried about this pattern from a security point of view. The idea behind UnpackSignedObject is that it is the boundary between the untrusted and trusted worlds, so we have a risk here that we check one thing then look at another.

Reply:
Hm, I don't understand, what do You mean

Reply:
Hi @patriotyk,  

The goal with `UnpackSignedObject()`  is to provide a clean, secure separation between metadata that is trusted, and untrusted metadata that has been downloaded from the internet. If we are not careful, then an attacker may be able to exploit a corner case to evade the signature checks.

The rule is that json blobs of the form `{'signatures': ... 'signed':...}` are untrusted, and once you are holding the contents of the signed portion, then you can consider it trusted. I'm sure @stevana will not be happy with the type system implications of this rule :)

`UnpackSignedObject` is intended to be the only gateway between these two worlds, and will only extract the `signed:` portion if signature validation is succeeds.  The reason I'm worried about this change is that is bridges these two worlds (that of  `{'signatures': ... 'signed':...}` and trusted metadata), and that will make it hard to audit for security.

Long term, @heartsucker has a nicer design here, where the validation is global, i.e. it takes an entire set of metadata and asks 'is this collection of metadata valid according to the Uptane rules', but for now I'd prefer if we can keep `UnpackSignedObject` as the only airlock between trusted and untrusted metadata.